### ISSUE TITLE
How to pass MicrosoftAppPassword to BotAuthentication attribute

### ISSUE THREAD

**Issue by maximilianlorent** [Microsoft/BotBuilder#989]
Hi everybody,

I want to create an Azure cloud service which will contain the logics for multiple different bots.
Thus for each bot I created one API controller which exposes the Messages endpoint.
In general everything is working fine.

However I'm working with Bot Framework V3 and as each endpoint belongs to a different bot and as in V3 each bot has a different MicrosoftAppId / MicrosoftAppPassword combination I'm asking myself how I can store these different credentials in my cloud service.

My first thought was just to add multiple MicrosoftAppID and MicrosoftAppPassword keys to the Web.config, such as 

```
<add key="MicrosoftAppId1" value="XXXXXXXXXXXXXXXXXXXXXXXXXX" />
<add key="MicrosoftAppPassword1" value="XXXXXXXXXXXXXXXXXXXX"/>
<add key="MicrosoftAppId2" value="XXXXXXXXXXXXXXXXXXXXXXXXXX"/>
<add key="MicrosoftAppPassword2" value="XXXXXXXXXXXXXXXXXXXX"/>
```

but as far as I understood this is not possible (or at least I don't know how to do the mapping between the bots and these keys then).

The other Idea I had was to use the [BotAuthentication] attribute and provide the credentials there, such as:

```
[BotAuthentication(MicrosoftAppId = "XXXXXXXXXX", MicrosoftAppPassword = "XXXXXXXXX")]
[HttpPost]
public virtual async Task<HttpResponseMessage> Post([FromBody] Activity activity)
```

In that case I could easily use different credentials for each API endpoint. But unfortunately the BotAuthentification attribute only provides the fields MicrosoftAppId, MicrosoftAppSettingName, OpenIdConfigurationUrl and DissableSelfIssuedTokens.
So there's no way of specifying a password.

So please can anybody help me out with this. I'm sure that this can be done somehow, but I do not see how.

Thanks a lot.

**Comment by willportnoy** [comment]
related issue: https://github.com/Microsoft/BotBuilder/issues/867

**Comment by Andrea-Orimoto** [comment]
@maximilianlorent did the related issue help? thanks.

**Comment by maximilianlorent** [comment]
@willportnoy and @Andrea-Orimoto unfortunately this does not help me right now.

In the related post @willportnoy wrote that one can pass the credentials - i.e. MicrosoftAppId and MicrosoftAppPassword - as parameters to BotAuthentification.
Unfortunately with version 3 of BotBuilder this seems not to be possible anymore. You can only pass MicrososftAppId, MicrosoftAppIdSeeintName, OpenIdConfigurationUrl and DisableSelfIssuedTokens.
So this solution unfortunately does not work for me.

Moreover it was propsed to work wih a dynamic ActionFilterAttribute, but as there was no concrete example, unfortunately I also don't know here how to proceed.

Nevertheless I made some progress in the last two weeks as I found out that my problem is even worse than I thought at the beginning.
In fact I have to solve two problems instead of one

1. Enable the Microsoft Bot Connector Service to authenticate to my cloud service in which the Message Controllers for multiple bots are hosted (here BotAuthentification comes into play and here I've got the problem that each bot has a different MicrosoftAppId and a different MicrosoftAppPassword)
2. Enable the bot logics within the cloud service to send messages back to the Bot Connector service.

For the first problem I still have no idea and it would be great if you can further support me here.

However for problem number two I made progress by using the hints of another post here (unfortunately I cannot find it again, so maybe someone else can add the ID if he or she knows).
In my bots I'm heavily using the BotBuilder Dialogs and I found out that using await `context.PostAsync `also seems to take it's credentials from the Web.config. However, as each of my bots uses different credentials storing the credentials in the Web.config is not a solution as only one pair of MicrosoftAppId and MicrosoftAppPassword can be stored there (feel free to correct me if I'm wrong). 
But fortunately here I found a solution which I modified a little bit and which works great for me. However, I'm not sure whether this is the best solution and whether this does not have any side impacts. So feel free to comment on this.

In each Message controller I call the following method before the typical `await Conversation.SendAsync(activity, () => new MainDialog(activity)) `(as can be seen in several examples on the Microsoft Bot Framework website).

```
public static void RegisterBotDependencies()
        {
            var builder = new ContainerBuilder();

            builder.Register(c => new BotIdResolver("BotId"))
                .As<IBotIdResolver>()
                .SingleInstance();

            builder.Register(c => new MicrosoftAppCredentials("MicrosoftAppId","MicrosoftAppPassword")).SingleInstance();

            builder.Update(Conversation.Container);
        }
```

So to sum it up:

1. I still need help with my BotAuthentication problem for multiple bots hosted in one cloud service
2. Tell me what you think about the RegisterBotDependencies solution.

Thanks a lot.

**Comment by willportnoy** [comment]
@maximilianlorent your RegisterBotDependencies isn't thread safe if you're updating a shared static container.

This is the particular comment I was referencing: https://github.com/Microsoft/BotBuilder/issues/867#issuecomment-237443188

ASP.NET has facilities for dynamically instantiating attributes from a dependency injection container - you might be able to select which attribute (and therefore which Microsoft app ID and password) to use based on data in the incoming request.  However, I have not tried this solution.

**Comment by dandriscoll** [comment]
Hi @maximilianlorent, I think there's some misunderstanding here.

Your Microsoft App Password is not required to verify calls coming in to the bot. That's why we don't expose a way to configure it.

**Comment by willportnoy** [comment]
@maximilianlorent I'm going to close this issue since @dandriscoll has provided more clarity on the question.  Feel free to open a new issue for further discussion.
